OpenRA.Mods.Common/Effects/LineImpactProjectile.cs

```diff
@@ -0,0 +1,262 @@
+				if (tailTicks >= length)
+				{
+					tailPos = target;
+					tailTravelling = false;
+				}
+				else
+					tailPos = WPos.LerpQuadratic(args.Source, target, angle, tailTicks, length);
+			}
+
+			// Raycast to find all affected actors
+			var hitActors = world.FindActorsOnLine(tailPos, headPos, info.BeamWidth, info.TargetExtraSearchRadius);
+
+			// If this projectile is blockable, first check if any of the hit actors will block this.
+			if (info.Blockable)
+			{
+				var wasBlocked = false;
+				foreach (var victim in hitActors)
+				{
+					if (victim.TraitsImplementing<BlocksProjectiles>().Any(Exts.IsTraitEnabled))
+					{
+						// Find where on the line this actor blocked the beam, and then limit the beam to that point.
+						var projectedPoint = FindActors.MinimumPointLineProjection(tailPos, headPos, victim.CenterPosition);
+						if ((projectedPoint - tailPos).Length < (headPos - tailPos).Length)
+						{
+							headPos = projectedPoint;
+							wasBlocked = true;
+						}
+					}
+				}
+
+				if (wasBlocked)
+				{
+					target = headPos;
+					length = Math.Min(headTicks, length);
+					hitActors = world.FindActorsOnLine(headPos, tailPos, info.BeamWidth, info.TargetExtraSearchRadius);
+				}
+			}
+
+			if (headTicks % info.TicksBetweenImpacts == 0)
+			{
+				foreach (var victim in hitActors)
+				{
+					var currentPos = victim.CenterPosition;
+					var distFromSource = (args.Source - currentPos).Length;
+					var distFromTarget = (target - currentPos).Length;
+
+					// Current falloff multiplier = 100 - ((Current distance - Falloff Start Range) / (Falloff End Range - Falloff Start Range) * Max Falloff);
+					// With the distance limited between 0 and falloff range.
+					var falloffRange = Math.Max((info.FalloffEndDistance - info.FalloffStartDistance).Length, 1);
+					var falloff = (currentPos - args.Source).Length - info.FalloffStartDistance.Length;
+					falloff = falloff.Clamp(0, falloffRange);
+					falloff = 100 - (falloff * info.FalloffMaxReduction / falloffRange);
+					falloff = falloff.Clamp(0, 100);
+
+					// Only allow impacts if far enough from source, and close enough to target.
+					if (distFromTarget < info.MaxTargetRange.Length)
+						Impact(victim, falloff);
+				}
+			}
+
+			var beamDone = !headTravelling && headTicks >= length && !tailTravelling && tailTicks >= length;
+			if (beamDone)
+				world.AddFrameEndTask(w => w.Remove(this));
+		}
+
+		public IEnumerable<IRenderable> Render(WorldRenderer wr)
+		{
+			if (wr.World.FogObscures(tailPos) &&
+				wr.World.FogObscures(headPos))
+				yield break;
+
+			var beamDone = !headTravelling && headTicks >= length && !tailTravelling && tailTicks >= length;
+
+			if (!beamDone)
+			{
+				yield return new BeamRenderable(headPos, 0, tailPos - headPos, info.BeamRenderWidth, color);
+			}
+		}
+
+		void Impact(Actor victim, int falloffFactor)
+		{
+			// Add the falloff factor as an additional damage modifier
+			var adjustedModifiers = new List<int>();
```

Comment:
```args.DamageModifiers``` is an array ```int[]```.
Not sure how to still allow ```adjustedModifiers.Add(falloffFactor);``` without first making it into a List.

However, after some searching, I've found ```adjustedModifiers.AddRange(args.DamageModifiers);``` to replace the for loop with.